### PR TITLE
fix(runtime): preload media on init to prevent broken first-play audio

### PR DESCRIPTION
**Pull request by miguel-heygen**
## Summary
- Audio (and video) sounds broken/choppy on first play in the studio preview, but works fine on second play
- Root cause: `<audio>` elements default to `preload="metadata"`, which only fetches enough data to determine duration — not enough for smooth playback. When `el.play()` fires, the browser hasn't buffered the audio data yet
- The runtime now eagerly sets `preload="auto"` and calls `load()` during init, ensuring media is fully buffered before the user clicks play
- `syncRuntimeMedia` now defers `play()` on unbuffered media by registering a `canplay` listener, instead of silently swallowing the failure

## Testing
Verified with agent-browser against a 26s narration composition (soulscape-film):

```
# After fix — audio element state at init:
preload: "auto"
readyState: 4 (HAVE_ENOUGH_DATA)
buffered: 26.07s (entire file)
duration: 26.07s
```

Audio is fully buffered before any play attempt, so first-play works identically to subsequent plays.

## Files changed
- `packages/core/src/runtime/init.ts` — set `preload="auto"` + `load()` in `bindMediaMetadataListeners`
- `packages/core/src/runtime/media.ts` — defer `play()` on unbuffered media via `canplay` listener
- `packages/core/src/runtime/media.test.ts` — updated test + added unbuffered media test case

### CODE REVIEW

File: packages/core/src/runtime/media.ts [hunk 1]
```diff
@@ -101,7 +101,19 @@ export function syncRuntimeMedia(params: {
...
+        } else {
+          // Media not yet buffered — start loading and play once ready.
+          if (el.preload !== "auto") el.preload = "auto";
+          const onCanPlay = () => {
+            el.removeEventListener("canplay", onCanPlay);
+            if (!el.paused) return; // already playing
+            void el.play().catch(() => {});
+          };
+          el.addEventListener("canplay", onCanPlay, { once: true });
+          el.load();
```
Comment (jrusso1020) [inline packages/core/src/runtime/media.ts:123]: 🔴 **Bug: `syncRuntimeMedia` is called every tick — this will restart `load()` on every frame.**

Per the HTML spec, `HTMLMediaElement.load()` **aborts** any in-progress fetch and restarts from zero. If `syncRuntimeMedia` runs at 60fps while the media is unbuffered, it calls `load()` 60×/sec, resetting the fetch each time. The media may never finish buffering — a regression worse than the original bug (choppy first-play → no play at all).

This also accumulates `canplay` listeners (each tick creates a new closure), all of which fire simultaneously when the media finally loads.

**Suggested fix**: Track elements with a pending deferred play (e.g. a module-level `WeakSet<HTMLMediaElement>`) and skip this branch if already pending:

```ts
const pendingPlay = new WeakSet<HTMLMediaElement>();

// inside syncRuntimeMedia:
if (el.readyState >= HTMLMediaElement.HAVE_FUTURE_DATA) {
  void el.play().catch(() => {});
} else if (!pendingPlay.has(el)) {
  pendingPlay.add(el);
  if (el.preload !== "auto") el.preload = "auto";
  el.addEventListener("canplay", () => {
    pendingPlay.delete(el);
    if (!el.paused) return;
    void el.play().catch(() => {});
  }, { once: true });
  el.load();
}
```

File: packages/core/src/runtime/media.ts [hunk 2]
```diff
@@ -101,7 +101,19 @@ export function syncRuntimeMedia(params: {
...
+        if (el.readyState >= HTMLMediaElement.HAVE_FUTURE_DATA) {
+          void el.play().catch(() => {});
+        } else {
+          // Media not yet buffered — start loading and play once ready.
+          if (el.preload !== "auto") el.preload = "auto";
+          const onCanPlay = () => {
+            el.removeEventListener("canplay", onCanPlay);
+            if (!el.paused) return; // already playing
+            void el.play().catch(() => {});
+          };
```
Comment (jrusso1020) [inline packages/core/src/runtime/media.ts]: 🟡 **Duplicate `load()` between init.ts and media.ts** — `init.ts` already calls `load()` once per element during `bindMediaMetadataListeners` (correctly guarded by the `metadataBoundMedia` Set). If that load is 90% done when the first `syncRuntimeMedia` tick hits this path, it throws away that progress and restarts.

With the `pendingPlay` guard suggested above, this becomes less of an issue since `load()` would only be called once here too. But worth noting the interaction.

File: packages/core/src/runtime/media.ts [hunk 3]
```diff
@@ -101,7 +101,19 @@ export function syncRuntimeMedia(params: {
...
-        void el.play().catch(() => {});
+        if (el.readyState >= HTMLMediaElement.HAVE_FUTURE_DATA) {
+          void el.play().catch(() => {});
+        } else {
+          // Media not yet buffered — start loading and play once ready.
+          if (el.preload !== "auto") el.preload = "auto";
+          const onCanPlay = () => {
+            el.removeEventListener("canplay", onCanPlay);
+            if (!el.paused) return; // already playing
+            void el.play().catch(() => {});
```
Comment (jrusso1020) [inline packages/core/src/runtime/media.ts]: 🟡 **No error/timeout for failed loads** — If the media 404s, has a CORS error, or the network drops, `canplay` never fires and the media silently stays in a "waiting to play" state forever. The old code at least attempted `play()`, which would reject (caught by `.catch()`).

Consider also listening for the `error` event alongside `canplay`, or adding a timeout fallback:
```ts
el.addEventListener("error", () => {
  pendingPlay.delete(el);
}, { once: true });
```

File: packages/core/src/runtime/media.ts [hunk 4]
```diff
@@ -101,7 +101,19 @@ export function syncRuntimeMedia(params: {
         }
       }
       if (params.playing && el.paused) {
-        void el.play().catch(() => {});
+        if (el.readyState >= HTMLMediaElement.HAVE_FUTURE_DATA) {
+          void el.play().catch(() => {});
+        } else {
+          // Media not yet buffered — start loading and play once ready.
+          if (el.preload !== "auto") el.preload = "auto";
+          const onCanPlay = () => {
```
Comment (jrusso1020) [inline packages/core/src/runtime/media.ts]: nit: The manual `el.removeEventListener("canplay", onCanPlay)` is redundant — `{ once: true }` already auto-removes the listener after the first invocation. Pick one pattern.

File: packages/core/src/runtime/media.test.ts [hunk 5]
```diff
@@ -171,12 +171,22 @@ describe("syncRuntimeMedia", () => {
...
   });
 
+  it("defers play on unbuffered media until canplay fires", () => {
+    const clip = createMockClip({ start: 0, end: 10 });
+    Object.defineProperty(clip.el, "readyState", { value: 0, writable: true });
+    const addEventSpy = vi.spyOn(clip.el, "addEventListener");
+    syncRuntimeMedia({ clips: [clip], timeSeconds: 5, playing: true, playbackRate: 1 });
+    expect(clip.el.play).not.toHaveBeenCalled();
+    expect(addEventSpy).toHaveBeenCalledWith("canplay", expect.any(Function), { once: true });
+  });
```
Comment (jrusso1020) [inline packages/core/src/runtime/media.test.ts:190]: 🟡 **Test gaps**:

1. **`load()` not verified** — This test asserts `play()` is NOT called and `addEventListener("canplay")` IS called, but never checks that `el.load()` was called. That's the trigger that makes the fix work.

2. **Callback not exercised** — The test never fires the `canplay` event, so the callback logic (`!el.paused` guard, the actual `play()` call) is untested.

3. **Repeated-call scenario missing** — The most critical edge case — `syncRuntimeMedia` called multiple times while unbuffered — has no test. This is the scenario that exposes the `load()` restart bug.

Suggested additions:
```ts
it("calls load() on unbuffered media", () => { ... });
it("plays when canplay fires", () => { /* dispatch canplay event, assert play() called */ });
it("does not re-register listener if already pending", () => { /* call twice, assert one listener */ });
```

**Review by jrusso1020 — DISMISSED** [review]
Nice root cause analysis and the `init.ts` change is clean. However, the `media.ts` change has a likely regression: `syncRuntimeMedia` is called every tick, so `el.load()` fires on every frame while the media is unbuffered — aborting and restarting the fetch each time, potentially preventing the media from ever finishing loading.

**Key issues:**
1. **🔴 Repeated `load()` in hot loop** — needs a guard (`WeakSet`) to skip if deferred play is already pending
2. **🟡 Duplicate `load()`** — `init.ts` already eagerly loads; `media.ts` can abort that in-progress fetch
3. **🟡 No error handling** — if media fails to load, `canplay` never fires, silent failure forever
4. **🟡 Test gaps** — `load()` call not verified, callback not exercised, repeated-call scenario untested

The `init.ts` change alone may be sufficient to fix the bug (it eagerly buffers during init). If the `media.ts` fallback is kept, it needs the `pendingPlay` guard — see inline suggestion.

File: packages/core/src/runtime/media.ts (hunk 1 above)
Comment (miguel-heygen) [reply packages/core/src/runtime/media.ts:123]: Fixed — added `pendingPlay` WeakSet guard. `load()` and `addEventListener` now fire exactly once per element.

File: packages/core/src/runtime/media.ts (hunk 2 above)
Comment (miguel-heygen) [reply packages/core/src/runtime/media.ts]: Good call — the `pendingPlay` guard now prevents the duplicate `load()`. The `init.ts` preload handles the eager case, and `media.ts` only triggers for elements that weren't caught during init (e.g. dynamically added media).

File: packages/core/src/runtime/media.ts (hunk 3 above)
Comment (miguel-heygen) [reply packages/core/src/runtime/media.ts]: Fixed — added `error` event listener alongside `canplay` that cleans up the `pendingPlay` entry on failure.

File: packages/core/src/runtime/media.ts (hunk 4 above)
Comment (miguel-heygen) [reply packages/core/src/runtime/media.ts]: Fixed — removed the redundant manual `removeEventListener`, using `{ once: true }` exclusively.

File: packages/core/src/runtime/media.test.ts (hunk 5 above)
Comment (miguel-heygen) [reply packages/core/src/runtime/media.test.ts:190]: All addressed: tests now verify `load()` is called, `canplay` callback fires and triggers `play()`, and repeated ticks don't re-register listeners.

**Review by jrusso1020 — APPROVED** [review]
